openfl-tutorials/interactive_api_tutorials_experimental/director_kvasir/envoy_folder/kvasir_shard_descriptor.py

```diff
@@ -0,0 +1,130 @@
+        # Settings for sharding the dataset
+        self.rank_worldsize = tuple(int(num) for num in rank_worldsize.split(','))
+
+        self.images_path = self.data_folder / 'segmented-images' / 'images'
+        self.masks_path = self.data_folder / 'segmented-images' / 'masks'
+
+        self.images_names = [img_name for img_name in sorted(os.listdir(
+            self.images_path)) if len(img_name) > 3 and img_name[-3:] == 'jpg']
+        # Sharding
+        self.images_names = self.images_names[self.rank_worldsize[0] - 1::self.rank_worldsize[1]]
+
+        # Calculating data and target shapes
+        sample, target = self[0]
+        self._sample_shape = [str(dim) for dim in sample.shape]
+        self._target_shape = [str(dim) for dim in target.shape]
+
+    @staticmethod
+    def download_data(data_folder):
+        """Download data."""
+        zip_file_path = data_folder / 'kvasir.zip'
+        os.makedirs(data_folder, exist_ok=True)
+        os.system('wget -nc'
+                  + " 'https://datasets.simula.no/hyper-kvasir/hyper-kvasir-segmented-images.zip'"
+                  + f' -O {zip_file_path.relative_to(Path.cwd())}')
+        zip_sha384 = 'e30d18a772c6520476e55b610a4db457237f151e' \
+                     '19182849d54b49ae24699881c1e18e0961f77642be900450ef8b22e7'
+        assert sha384(open(zip_file_path, 'rb').read(
+            os.path.getsize(zip_file_path))).hexdigest() == zip_sha384
+        os.system(f'unzip -n {zip_file_path.relative_to(Path.cwd())}'
+                  + f' -d {data_folder.relative_to(Path.cwd())}')
+
+    def __getitem__(self, index):
+        """Return a item by the index."""
+        name = self.images_names[index]
+        # Reading data
+        img = Image.open(self.images_path / name)
+        mask = Image.open(self.masks_path / name)
+        if self.enforce_image_hw is not None:
+            # If we need to resize data
+            # PIL accepts (w,h) tuple, not (h,w)
+            img = img.resize(self.enforce_image_hw[::-1])
+            mask = mask.resize(self.enforce_image_hw[::-1])
+        img = np.asarray(img)
+        mask = np.asarray(mask)
+        assert img.shape[2] == 3
+
+        return img, mask[:, :, 0].astype(np.uint8)
+
+    def __len__(self):
+        """Return the len of the dataset."""
+        return len(self.images_names)
+
+    @property
+    def sample_shape(self):
+        """Return the sample shape info."""
+        return self._sample_shape
+
+    @property
+    def target_shape(self):
+        """Return the target shape info."""
+        return self._target_shape
+
+    @property
+    def dataset_description(self) -> str:
+        """Return the dataset description."""
+        return f'Kvasir dataset, shard number {self.rank_worldsize[0]}' \
+               f' out of {self.rank_worldsize[1]}'
+
+
+if __name__ == '__main__':
+    from openfl.interface.cli import setup_logging
+    setup_logging()
+
+    data_folder = 'data'
+    rank_worldsize = '1,100'
```

Comment:
The envoy would not necessarily know ahead of time what the rank and world size are. Ideally, the rank and world size should be determined by the Director, and then you could trigger a delayed_init function (similar to interactive API dataset definition) to perform the actual sharding at the start of the experiment (based on envoys registered with director)

Reply:
There is ambiguity in this example, envoy may be started just from the shard descriptors `main` function. But the primary way is to write a [plan-like yaml file with a shard descriptor path](https://github.com/aleksandr-mokrov/openfl/blob/director-envoy/openfl-tutorials/interactive_api_tutorials_experimental/director_kvasir/envoy_folder/shard_config.yaml) and required parameters.
Parameters may or may not contain rank and worldsize, probably those make sense only for artificial federations.
In general, a shard descriptor should be written by the shard owner and may use arbitrary parameters.
For the `_delayed_init` - it is not a part of the interface anymore. Now, as we need to adapt local shard descriptor on collaborator node, we ask user to implement shard descriptor setter in their FederatedDataset class.

Reply:
The automatic rank/world assignment doesn't need to be solved for this PR, but this is an area that should be improved going forward